File: ios/alier_dev/alier/http/data/ResponseData.swift
/*
Copyright 2024 Suredesigns Corp.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import Foundation

class ResponseData {
    var status: Int
    var statusText: String
    var headers: Headers
    var body: ByteArray
    
    init(status: Int, statusText: String, headers: Headers, body: ByteArray?) {
        self.status = status
        self.statusText = statusText
        self.headers = headers
        self.body = body == nil ? ByteArray("".utf8) : body!
    }
    
    func equals(other: Any?) -> Bool {
        if self === other as! ResponseData { return true }
        if other is ResponseData {
            let otherData = other as! ResponseData
            if self.status != otherData.status { return false }
            if self.statusText != otherData.statusText { return false }
            if !self.headers.equals(other: otherData.headers)  { return false }
            if self.body.isEmpty || otherData.body.isEmpty  { return self.body == otherData.body }

            return self.body == otherData.body
        } else {
            return false
        }
    }

    func toDict() -> Dictionary<String, Any?> {
        let content_type = (self.headers.get(key: "content-type") != nil) ? self.headers.get(key: "content-type") : "application/octet-stream"
        let encStrBody: String?
        if self.body.isEmpty {
            encStrBody = nil
        } else {
            let data = Data(self.body)
            let encoded = data.base64EncodedString()
            encStrBody = "data:\(content_type!);base64,\(encoded)"
        }
        let dict : [String: Any?] = [
            "status"     : self.status,
            "statusText" : self.statusText,
            "body"       : encStrBody,
            "headers"    : self.headers.toDict()
        ]
        return dict
    }
}

File: android/alier_dev/alier/src/main/java/com/suredesigns/alier/net/http/data/ResponseData.kt
/*
Copyright 2024 Suredesigns Corp.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.suredesigns.alier.net.http.data

import android.util.Base64

data class ResponseData(val status: Int, val statusText: String, val headers: Headers, val body: ByteArray?) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ResponseData) return false

        if (status != other.status) return false
        if (statusText != other.statusText) return false
        if (headers != other.headers) return false
        if ((body === null) or (other.body === null)) return body === other.body

        return body.contentEquals(other.body)
    }

    fun toMap(): Map<String, Any?> {
        val content_type = headers["content-type"] ?: "application/octet-stream"
        return mapOf(
            "status" to status,
            "statusText" to statusText,
            "body" to if (body == null || body.isEmpty()) null else {
                "data:${content_type};base64,${Base64.encodeToString(body, Base64.NO_WRAP + Base64.URL_SAFE)}"
            },
            "headers" to headers.toMap()
        )
    }

    override fun hashCode(): Int {
        var result = status
        result = 31 * result + statusText.hashCode()
        result = 31 * result + headers.hashCode()
        if (body != null) {
            result = 31 * result + body.contentHashCode()
        }
        return result
    }
}